### ISSUE TITLE
No support for client certificate chains. urllib3 1.19.1

### ISSUE THREAD

**Issue by kevgliss** [shazow/urllib3#1060]
Ran into an issue with client certificates where chain validation is failing.

Orginally reported:
https://github.com/kennethreitz/requests/issues/3732

Traceback:
```
Traceback (most recent call last):
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/contrib/pyopenssl.py", line 417, in wrap_socket
    cnx.do_handshake()
  File "/apps/lemur/lib/python3.5/site-packages/OpenSSL/SSL.py", line 1424, in do_handshake
    self._raise_ssl_error(self._ssl, result)
  File "/apps/lemur/lib/python3.5/site-packages/OpenSSL/SSL.py", line 1172, in _raise_ssl_error
    _raise_current_error()
  File "/apps/lemur/lib/python3.5/site-packages/OpenSSL/_util.py", line 48, in exception_from_error_queue
    raise exception_type(errors)
OpenSSL.SSL.Error: [('SSL routines', 'SSL3_READ_BYTES', 'tlsv1 alert unknown ca')]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 594, in urlopen
    chunked=chunked)
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 350, in _make_request
    self._validate_conn(conn)
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 835, in _validate_conn
    conn.connect()
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/connection.py", line 323, in connect
    ssl_context=context)
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/util/ssl_.py", line 324, in ssl_wrap_socket
    return context.wrap_socket(sock, server_hostname=server_hostname)
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/contrib/pyopenssl.py", line 424, in wrap_socket
    raise ssl.SSLError('bad handshake: %r' % e)
ssl.SSLError: ("bad handshake: Error([('SSL routines', 'SSL3_READ_BYTES', 'tlsv1 alert unknown ca')],)",)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/apps/lemur/lib/python3.5/site-packages/requests/adapters.py", line 423, in send
    timeout=timeout
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 624, in urlopen
    raise SSLError(e)
requests.packages.urllib3.exceptions.SSLError: ("bad handshake: Error([('SSL routines', 'SSL3_READ_BYTES', 'tlsv1 alert unknown ca')],)",)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/apps/lemur/lemur/common/schema.py", line 147, in decorated_function
    resp = f(*args, **kwargs)
  File "/apps/lemur/lemur/certificates/views.py", line 271, in post
    return service.create(**data)
  File "/apps/lemur/lemur/certificates/service.py", line 218, in create
    cert_body, private_key, cert_chain = mint(**kwargs)
  File "/apps/lemur/lemur/certificates/service.py", line 167, in mint
    cert_body, cert_chain = issuer.create_certificate(csr, kwargs)
  File "/apps/lemur/lemur-cloudca/lemur_cloudca/plugin.py", line 332, in create_certificate
    response = self.post(endpoint, cloudca_options)
  File "/apps/lemur/lemur-cloudca/lemur_cloudca/plugin.py", line 163, in post
    response = self.session.post(self.url + endpoint, data=dumps(data), timeout=10, verify=self.ca_bundle)
  File "/apps/lemur/lib/python3.5/site-packages/requests/sessions.py", line 535, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/apps/lemur/lib/python3.5/site-packages/requests/sessions.py", line 488, in request
    resp = self.send(prep, **send_kwargs)
  File "/apps/lemur/lib/python3.5/site-packages/requests/sessions.py", line 609, in send
    r = adapter.send(request, **kwargs)
  File "/apps/lemur/lib/python3.5/site-packages/requests/adapters.py", line 497, in send
    raise SSLError(e, request=request)
requests.exceptions.SSLError: ("bad handshake: Error([('SSL routines', 'SSL3_READ_BYTES', 'tlsv1 alert unknown ca')],)",)

Traceback (most recent call last):
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 594, in urlopen
    chunked=chunked)
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 350, in _make_request
    self._validate_conn(conn)
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 835, in _validate_conn
    conn.connect()
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/connection.py", line 323, in connect
    ssl_context=context)
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/util/ssl_.py", line 324, in ssl_wrap_socket
    return context.wrap_socket(sock, server_hostname=server_hostname)
  File "/apps/lemur/lib/python3.5/site-packages/requests/packages/urllib3/contrib/pyopenssl.py", line 424, in wrap_socket
    raise ssl.SSLError('bad handshake: %r' % e)
ssl.SSLError: ("bad handshake: Error([('SSL routines', 'SSL3_READ_BYTES', 'tlsv1 alert unknown ca')],)",)
```
Reproduce with:

```
import requests

class HostNameCheckingAdapter(HTTPAdapter):
    def cert_verify(self, conn, url, verify, cert):
        super(HostNameCheckingAdapter, self).cert_verify(conn, url, verify, cert)
        conn.assert_hostname = False

s = requests.Session()
s.mount('https://', HostNameCheckingAdapter())
s.cert = ('client.crt', 'client.key')

s.get('https://example.com', verify='path-to-ca-bundle')
```

Per @Lukasa comment, modifying the following line resolves the issue:
pyopenssl.py

    self._ctx.use_certificate_file(certfile)

to

    self._ctx.use_certificate_chain_file(certfile)

**Comment by Lukasa** [comment]
Cool, so this is a really easy fix. The hardest part is making sure we add a regression test to avoid it happening again. I'm going to put this up for contributors to handle.

**Comment by SethMichaelLarson** [comment]
I'll pick this up tonight. :)

**Comment by SethMichaelLarson** [comment]
Related: #878

**Comment by Lukasa** [comment]
Oh great spot! Let's squash both of these then.

**Comment by SethMichaelLarson** [comment]
So I'm having trouble creating a client certificate chain, I'm not a master of OpenSSL so I doubt I should be the one creating this certificates in case I'm doing something incorrectly. Maybe I shouldn't be the one to pick up this issue. :(

**Comment by chet** [comment]
Hey @SethMichaelLarson! Just wanted to drop a note that I ran across this problem recently in my development efforts, and was eventually led to this issue, which the above one liner fixed. :)

In my case, I was using `requests` w/ a client certificate, and everything worked fine, but once I installed `pyopenssl`, kaboom, it all broke. It took a while for me to track it down, but I was excited when I did; the change above fixed everything, so, thanks @Lukasa!

That said, I'd prefer to not carry a patch, so, what sort of help would you need here to make sure you're comfortable with the fix?

If it helps, https://jamielinux.com/docs/openssl-certificate-authority/ is a good HOWTO that I know a few people have followed to set up their own root CA. Once you do that, you'll need to also create an intermediate CA signed by your root CA, then issue a certificate off of that.

You could also get a certificate for free from https://letsencrypt.org, and use the signing authority's chain to verify the fix.

Just trying to brainstorm a few things. :)

**Comment by Lukasa** [comment]
I mean, in general we don't need any of those things. We already have a CA floating around in `dummyserver/certs`, all we need is to create an intermediate CA and then a leaf that it has signed. Shouldn't be too hard for those who are comfortable with OpenSSL.

**Comment by theonewolf** [comment]
@lukasa you mostly want contributor help in creating the test case?

Just so I'm clear, we can do a one line change and then add a test case and that would basically be merge ready (after review of course)?

**Comment by theonewolf** [comment]
Also @Lukasa just as an FYI (@kennethreitz) this causes `requests > 2.11.0` API to break.

Here is a simple snippet of what I did to track down why `requests` is now broken (from before) if upgrading past `2.11.0`:

```bash
> git clone https://github.com/kennethreitz/requests.git
> cd requests
> git bisect start
> git bisect bad
> git bisect good v2.11.0
> git bisect run python3 -c "from requests import __version__, get; print(__version__); exit(get('https://127.0.0.1', cert=('test.crt','test.key'), verify='root.crt').status_code != 200)"
```

where `test.crt` is two PEM-encoded certificates concatenated together (leaf + intermediate), and the result is:

```bash
99fa7becf263473c7bfc1998b41c2c6c80a0f499 is the first bad commit
commit 99fa7becf263473c7bfc1998b41c2c6c80a0f499
Author: Cory Benfield <lukasaoz@gmail.com>
Date:   Tue Nov 15 10:09:11 2016 +0000

    Update urllib3 to 1.19

:040000 040000 3d52e7d62b52eed4b72f81e113c31769ac886ee1 5090180050cc2e986a4f20b62beef37cce8f470e M	requests
bisect run success
```

**Comment by theonewolf** [comment]
@Lukasa ah!  You already know the offending commit :laughing: 

Guess we just need a merge request here :smile:

**Comment by Lukasa** [comment]
@theonewolf Yup, this is a really nice straightforward patch for someone who wants it!

**Comment by theonewolf** [comment]
@Lukasa do I need to do anything special to have PyOpenSSL tested?

Here's what I've done so far:

1. Created an intermediate CA
2. Created a client certificate off of that intermediate
3. Modified `dummyserver/server.py` to take certificates with `CERT_OPTIONAL`
    - Pre-existing test cases run unchanged
    - Test cases presenting validating certificates pass that along into the request handlers (Tornado)
3. Created two new test cases in `test/with_dummyserver/test_https.py`
    - One that tries a connection with client certificate alone
    - One that tries a connection with intermediate appended to client certificate

It seems to work perfectly!  But, I am worried that the test cases are defaulting to using the `ssl` built in library?

I run this to isolate my test cases:

```
nosetests test.with_dummyserver.test_https:TestHTTPS.test_client_intermediate
nosetests test.with_dummyserver.test_https:TestHTTPS.test_client_no_intermediate
```

**Comment by theonewolf** [comment]
@Lukasa disregard, running `tox` or `make test-all` successfully bubbles up the PyOpenSSL error :smile:.

**Comment by theonewolf** [comment]
This is what it looks like when it fails:

```
======================================================================
ERROR: test_client_intermediate (test.contrib.test_pyopenssl.transplant_class.<locals>.C)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/wolf/Desktop/urllib3/test/with_dummyserver/test_https.py", line 79, in test_client_intermediate
    r = https_pool.request('GET', '/certificate')
  File "/home/wolf/Desktop/urllib3/urllib3/request.py", line 66, in request
    **urlopen_kw)
  File "/home/wolf/Desktop/urllib3/urllib3/request.py", line 87, in request_encode_url
    return self.urlopen(method, url, **extra_kw)
  File "/home/wolf/Desktop/urllib3/urllib3/connectionpool.py", line 630, in urlopen
    raise SSLError(e)
urllib3.exceptions.SSLError: ("bad handshake: Error([('SSL routines', 'ssl3_read_bytes', 'tlsv1 alert unknown ca')],)",)
-------------------- >> begin captured logging << --------------------
urllib3.connectionpool: DEBUG: Starting new HTTPS connection (1): localhost
tornado.general: WARNING: SSL Error on 8 ('127.0.0.1', 58512): [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:645)
--------------------- >> end captured logging << ---------------------
```

**Comment by theonewolf** [comment]
@Lukasa pull request in :smile:.

**Comment by andamian** [comment]
Are there any plans for releasing this any time soon? As a work around to it, is it possible to make urllib3 used by requests not use pyOpenSSL even if it's installed? Thanks

**Comment by Lukasa** [comment]
There is: run `requests.packages.urllib3.contrib.pyopenssl.extract_from_urllib3`.

**Comment by SethMichaelLarson** [comment]
This was merged in #1246
